File: sample_race.py
"""Python implementation of race condition for testing"""
import threading


class UnsafeCounter:
    def __init__(self):
        self.value = 0
        self.lock = threading.Lock()  # Will be instrumented

    def increment(self):
        # Simulate unsafe access pattern
        self.value += 1


class SafeCounter:
    def __init__(self):
        self.value = 0
        self.lock = threading.Lock()

    def increment(self):
        with self.lock:
            self.value += 1


def run_counter_test(counter_class, threads=5, iterations=10000):
    counter = counter_class()
    threads = []

    def worker():
        for _ in range(iterations):
            counter.increment()

    for _ in range(threads):
        t = threading.Thread(target=worker)
        t.start()
        threads.append(t)

    for t in threads:
        t.join()

    return counter.value

File: tests/test_data/sample_race.cpp
/*
 * Sample test file with a data race for testing ThreadGuard.
 * This file contains a simple data race between two threads.
 */

#include <thread>
#include <iostream>

int counter = 0;

void increment() {
    for (int i = 0; i < 1000; ++i) {
        counter++;  // Data race here
    }
}

int main() {
    std::thread t1(increment);
    std::thread t2(increment);

    t1.join();
    t2.join();

    std::cout << "Final counter value: " << counter << std::endl;
    return 0;
}
